sound/hda/hdac_device.c

```diff
@@ -109,12 +109,16 @@ int snd_hdac_device_init(struct hdac_device *codec, struct hdac_bus *bus,
 				     codec->vendor_id & 0xffff);
 	if (!codec->chip_name) {
 		err = -ENOMEM;
-		goto error;
+		goto error_chip;
 	}
 
 	return 0;
 
+ error_chip:
+	kfree(codec->vendor_name);
  error:
+	pm_runtime_put_noidle(&codec->dev);
+	pm_runtime_set_suspended(&codec->dev);
```

Comment:
@RanderWang This is a nice catch. But Im wondering if it makes sense to move the following to the end of the init() function instead of doing it at the beginning and rolling back in case of errors.
```
pm_runtime_set_active(&codec->dev);
pm_runtime_get_noresume(&codec->dev);
atomic_set(&codec->in_pm, 0);
```

Reply:
@ranj063  please check the following code. snd_hdac_ext_bus_device_exit will be called if snd_hdac_device_register is failed. so it is better to do it at snd_hdac_ext_bus_device_exit

````
	ret  = snd_hdac_device_init(hdev, bus, name, addr);
	if (ret < 0) {
		dev_err(bus->dev, "device init failed for hdac device\n");
		return ret;
	}
	hdev->type = HDA_DEV_ASOC;
	hdev->dev.release = default_release;

	ret = snd_hdac_device_register(hdev);
	if (ret) {
		dev_err(bus->dev, "failed to register hdac device\n");
		snd_hdac_ext_bus_device_exit(hdev);
		return ret;
	}
````

Reply:
@RanderWang all I meant was to move those 3 lines just before return 0 in the snd_hdac_device_init() function. Of course, the exit function needs to do it as well.

Reply:
anyway its only a minor improvement. I'm good with this change too